core/server/master/src/main/java/alluxio/master/meta/MetaDailyBackup.java

```diff
@@ -0,0 +1,182 @@
+
+    mBackup = mScheduleExecutor.scheduleAtFixedRate(new Runnable() {
+          @Override
+          public void run() {
+            dailyBackup();
+          }
+        }, getTimeToNextBackup(), FormatUtils.parseTimeSize("1day"),
+        TimeUnit.MILLISECONDS);
+    LOG.info("MetaDailyBackup scheduled.");
+  }
+
+  /**
+   * Gets the time gap between now and next backup time.
+   *
+   * @return the time gap to next backup
+   */
+  private long getTimeToNextBackup() {
+    LocalDateTime now = LocalDateTime.now(Clock.systemUTC());
+    String[] hourAndMin = Configuration.get(PropertyKey.MASTER_DAILY_BACKUP_TIME).split(":");
+    int hour = Integer.parseInt(hourAndMin[0]);
+    int min = hourAndMin.length == 2 ? Integer.parseInt(hourAndMin[1]) : 0;
+    LocalDateTime nextBackupTime = LocalDateTime.of(now.getYear(),
+        now.getMonth(), now.getDayOfMonth(), hour, min, 0);
+    if (nextBackupTime.isBefore(now)) {
+      nextBackupTime = nextBackupTime.plusDays(1);
+    }
+    return ChronoUnit.MILLIS.between(now, nextBackupTime);
+  }
+
+  /**
+   * The daily backup task.
+   */
+  private void dailyBackup() {
+    String dir = Configuration.get(PropertyKey.MASTER_BACKUP_DIRECTORY);
+    String rootUfs = Configuration.get(PropertyKey.MASTER_MOUNT_TABLE_ROOT_UFS);
+    boolean isLocal = URIUtils.isLocalFilesystem(Configuration
+        .get(PropertyKey.MASTER_MOUNT_TABLE_ROOT_UFS));
+    if (isLocal) {
+      dir = PathUtils.concatPath(rootUfs, dir);
+    }
+    try {
+      BackupResponse resp = mMetaMaster.backup(new BackupOptions(dir, isLocal));
+      LOG.info("Successfully backed up journal to {}", resp.getBackupUri());
+      try {
+        deleteOldestBackups(dir, isLocal);
+      } catch (Exception e) {
+        LOG.error("Failed to delete outdated backup files at {}", dir, e);
+      }
+    } catch (Exception e) {
+      LOG.error("Failed to execute daily backup at {}", dir, e);
+    }
+  }
+
+  /**
+   * Deletes oldest backup files to avoid consuming too many spaces.
+   *
+   * @param dir the backup directory
+   * @param isLocal whether the root ufs is local filesystem
+   */
+  private void deleteOldestBackups(String dir, boolean isLocal) throws Exception {
+    UnderFileSystem ufs;
+    if (isLocal) {
+      ufs = UnderFileSystem.Factory.create("/", UnderFileSystemConfiguration.defaults());
+    } else {
+      ufs = UnderFileSystem.Factory.createForRoot();
+    }
+
+    UfsStatus[] statues = ufs.listStatus(dir);
+    if (statues.length <= mMaxFile) {
+      return;
+    }
+
+    // Sort the backup files according to create date from oldest to newest
+    TreeMap<LocalDateTime, String> dateToFile = new TreeMap<>((a, b) -> (
+        a.isBefore(b) ? -1 : a.isAfter(b) ? 1 : 0));
+    for (UfsStatus status : statues) {
+      if (status.isFile()) {
+        Matcher matcher = mBackupPattern.matcher(status.getName());
+        if (matcher.matches()) {
+          LocalDateTime date = LocalDateTime.ofInstant(Instant
```

Comment:
is there an advantage to sorting by `LocalDateTime` instead of `Instant`?

Reply:
Changed to sorting by  `Instant` to make the whole logic simpler. It's great than `Instant` also support `isBefore` and `isAfter` methods.